packages/react-app/src/App.jsx

```diff
@@ -91,20 +170,38 @@ function App({ mainnet, localProvider, appChainId }) {
       // metamask not installed
       return
     }
-    const provider = window.ethereum
-    // @ts-ignore
-    setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
+    const provider = appState.provider // window.ethereum
+    await provider.send('eth_requestAccounts', [])
 
-    provider.on('chainChanged', chainId => {
-      console.log(`chain changed to ${chainId}! updating providers`)
-      // @ts-ignore
-      setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
+    /**
+     * @param accountPayload string[]
+     */
+    provider.on('accountsChanged', async accountPayload => {
+      // accountPayload Array<string>
+      console.log(`account changed!`)
+      if (accountPayload.length === 0) {
+        console.log('Metamask requires login or no accounts added')
+        // show toast if need be
+      } else if (accountPayload[0] !== connectedAddress) {
+        setConnectedAddress(accountPayload[0])
+      }
     })
 
-    provider.on('accountsChanged', () => {
-      console.log(`account changed!`)
-      // @ts-ignore
-      setInjectedProvider(new ethers.providers.Web3Provider(window.ethereum))
+    provider.on('chainChanged', chainId => {
+      console.log(`chain changed to ${chainId}! updating providers`)
+      if (chainId === 5 || chainId === '5') {
+        // @ts-ignore
+        appDispatch({ actionType: APPSTATEACTION.GOERLICHAINID })
+        window.location.reload()
+      }
+      if (chainId === 10 || chainId === '10') {
+        // @ts-ignore
+        appDispatch({ actionType: APPSTATEACTION.OPTIMISMCHAINID })
+        window.location.reload()
+      }
+      if (chainId !== 10 || chainId !== '10' || chainId !== 5 || chainId !== '5') {
```

Comment:
there are some duplicate

Reply:
I have since removed these lines of code. Thanks